libflux/src/semantic/parser/mod.rs

```diff
@@ -0,0 +1,1696 @@
+use std::{ str::Chars, iter::Peekable, slice::Iter, collections::{ HashMap, HashSet } };
+
+use crate::semantic::types::{ PolyType, MonoType, Tvar, Kind, TvarKinds, Row, Array, Function, Property };
+
+#[derive(Debug, PartialEq, Copy, Clone)]
+// TokenTypes holds all possible TokenType values used by the Lexer and Parser. 
+pub enum TokenTypes {
+    ERROR,
+    EOF, 
+
+    // Keywords and Kinds
+    FORALL,
+    WHERE,
+    IS, // not currently in use for Kinds; use ':' instead
+    AND, 
+
+    // Literals and Identifiers
+    INT,
+    UINT,
+    FLOAT,
+    STRING,
+    BOOL,
+    DURATION, 
+    TIME,
+    REGEX,
+    IDENTIFIER,
+    WHITESPACE,
+
+    // Operators
+    LEFTCURLYBRAC,
+    RIGHTCURLYBRAC,
+    LEFTSQUAREBRAC,
+    RIGHTSQUAREBRAC,
+    LEFTPAREN,
+    RIGHTPAREN,
+    QUESTIONMARK,
+    COLON, 
+    COMMA, 
+    OR, 
+    PIPE, 
+    ARROW, 
+    PLUS
+}
+
+/* LEXER */
+// Lexer holds the iterator for the source string, the list of output tokens and keeps track of
+// the current string value of any identifier or keyword to append the the Token.
+struct Lexer<'a> {
+    source: Peekable<Chars<'a>>, 
+    tokens: Vec<Token>, 
+    current_string: String,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+// Token hold the token's type and the text value for any keyword or identifier to be used later 
+// by the parser. 
+pub struct Token {
+    token_type: TokenTypes,
+    text: Option<String>,
+}
+
+// Lex instantiates the Lexer with default values and initializes lexing. 
+// This function is not meant to be used directly. The user should pass 
+// source into Parse(), which in turn calls this function. 
+pub fn Lex(source: &str) -> Vec<Token> {
+    let mut lexer = Lexer { 
+        source: source.chars().peekable(),
+        tokens: Vec::new(),
+        current_string: String::new(),
+        };
+    lexer.lex_tokens(); 
+    lexer.tokens
+}
+
+// StateFn maintains the lexers recursive state
```

Comment:
What recursive state does the lexer have? I'm not too familiar with the language, but lexers shouldn't have recursive state. A parser may have recursive state, but most lexers don't. I'm looking and I don't see any location where `StateFn` is set to anything other than `lex_token`.

Reply:
`lex_token()` and `StateFn()` maintain the lexing behavior by calling one another while `Lexer.next()` iterates through the characters of the `source` string. I may be incorrect to call this behavior recursive, but that is how it appeared to me.